src/pamiq_core/torch/trainer.py

```diff
@@ -0,0 +1,224 @@
+        self.optimizers.clear()
+        self.lr_schedulers.clear()
+
+        # Process configuration based on return type
+        if isinstance(optimizer_config, tuple):
+            # Configuration includes both optimizers and schedulers
+            optimizers, schedulers = optimizer_config
+            self.optimizers.update(optimizers)
+            self.lr_schedulers.update(schedulers)
+        else:
+            # Configuration includes only optimizers
+            self.optimizers.update(optimizer_config)
+
+        # Restore optimizer states if available
+        for name, state in self.optimizer_states.items():
+            self.optimizers[name].load_state_dict(state)
+        # Restore scheduler states if available
+        for name, state in self.lr_scheduler_states.items():
+            self.lr_schedulers[name].load_state_dict(state)
+
+    def _keep_optimizer_and_scheduler_states(self) -> None:
+        """Keep the current states of optimizers and schedulers.
+
+        Captures and stores the state dictionaries of all active
+        optimizers and schedulers for future restoration.
+        """
+        # Clear previous kept states
+        self.optimizer_states.clear()
+        self.lr_scheduler_states.clear()
+
+        # Keep current optimizer states
+        for name, optimizer in self.optimizers.items():
+            self.optimizer_states[name] = optimizer.state_dict().copy()
+
+        # Keep current scheduler states
+        for name, scheduler in self.lr_schedulers.items():
+            self.lr_scheduler_states[name] = scheduler.state_dict().copy()
+
+    @override
+    def setup(self) -> None:
+        """Set up the training environment.
+
+        Initializes optimizers and schedulers by calling the `configure_optimizers`
+        method and restores their states if previously saved.
+        """
+        super().setup()
+        self._setup_optimizers_and_schedulers()
+
+    @override
+    def teardown(self) -> None:
+        """Clean up after training.
+
+        Keeps the current state of optimizers and schedulers before
+        cleanup.
+        """
+        super().teardown()
+        self._keep_optimizer_and_scheduler_states()
+
+    @override
+    def save_state(self, path: Path) -> None:
+        """Save trainer state to disk.
+
+        Persists the states of all optimizers and schedulers to the specified
+        directory path.
+
+        Args:
+            path: Directory path where state should be saved
+        """
+        super().save_state(path)
+
+        # Save optimizer states to disk
+        for name, optimizer_state in self.optimizer_states.items():
+            torch.save(optimizer_state, path / f"{name}.optim.pt")  # pyright: ignore[reportUnknownMemberType]
+
+        # Save scheduler states to disk
+        for name, scheduler_state in self.lr_scheduler_states.items():
+            torch.save(scheduler_state, path / f"{name}.lrsch.pt")  # pyright: ignore[reportUnknownMemberType]
```

Comment:
一度`teardown`を実行しないと`save_state`で正しく保存できないのなんか気になる　`teardown`ってモデル全部の情報を消去するようなニュアンスがあるけど、この実装だとその後に`save_state`ができてしまうし、なんならそうしなければならないの非自明だと思った
関数の頭で`self._keep_optimizer_and_scheduler_states()`を実行しておくといいかも？
```suggestion
    def save_state(self, path: Path) -> None:
        """Save trainer state to disk.

        Persists the states of all optimizers and schedulers to the specified
        directory path.

        Args:
            path: Directory path where state should be saved
        """
        super().save_state(path)
        
        # Prepare parameters to save
        self._keep_optimizer_and_scheduler_states()

        # Save optimizer states to disk
        for name, optimizer_state in self.optimizer_states.items():
            torch.save(optimizer_state, path / f"{name}.optim.pt")  # pyright: ignore[reportUnknownMemberType]

        # Save scheduler states to disk
        for name, scheduler_state in self.lr_scheduler_states.items():
            torch.save(scheduler_state, path / f"{name}.lrsch.pt")  # pyright: ignore[reportUnknownMemberType]
```

Reply:
学習後の状態が存在しない場合は不要な保存処理になるので無駄。最小のコードで必要な機能を実現している。

Reply:
これも上に同じく、もしこれで行くならコメントが欲しい

Reply:
コメント追加 https://github.com/MLShukai/pamiq-core/pull/177/commits/8eabd140a69b1fc3b54af9b1087e7c7954d23b95